### PR TITLE
Performance improvement: ReactDOMServer `resolve` class method

### PR DESCRIPTION
**Pull request by overlookmotel**
At present the server-side renderer has a hot path through a function called `resolve()` which contains a lengthy closure `processChild()`.

This PR converts `resolve()` and `processChild()` to prototype methods of `ReactDOMServerRenderer` class, to remove this closure.

Using a [fairly naive benchmark](https://github.com/overlookmotel/reactdom-server-bench), I am seeing the following performance improvements:

* Node 11: ~4%
* Node 10: ~4%
* Node 8: ~4%
* Node 6: ~15%

If there is an official benchmark I can use to test this on, please let me know.

### CODE REVIEW

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 1]
```diff
@@ -1200,6 +928,286 @@ class ReactDOMServerRenderer {
...
+      if (__DEV__) {
+        pushElementToDebugStack(element);
+      }
+      if (typeof Component !== 'function') {
+        break;
+      }
+      ({child, context} = this.resolveNext(element, context, Component));
+    }
+
+    return {child, context};
```
Comment (gaearon) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: Can we just return the thing returned from `resolveNext` here? Since we already have it.

**Comment by jacobp100** [comment]
Could this not be made simpler by just un-nesting `processChild` and passing in `threadID` as an argument? Then we could avoid the prototype lookup and the closure.

**Comment by gaearon** [comment]
Yeah I'd like to avoid dynamic dispatch if possible.

**Comment by overlookmotel** [comment]
Is there any official (or more official than mine) benchmark I can run to determine what's fastest?

**Comment by overlookmotel** [comment]
Please do not close, Mr Robot.

**Comment by gaearon** [comment]
>Is there any official (or more official than mine) benchmark I can run to determine what's fastest?

Not really. We have https://github.com/facebook/react/tree/master/scripts/bench but it's pretty outdated and doesn't server render.

**Comment by gaearon** [comment]
Could you try this please:

>Could this be made simpler by just un-nesting processChild and passing in threadID as an argument? Then we could avoid the prototype lookup and the closure.

?

**Pull request closed without merging**